File: Irrgarten-Java/src/irrgarten/Player.java
/**
 * @file irrgarten/Player.java
 * @author Álvaro Maldonado Medina, grupo: 2D-D3
 */

package irrgarten;

import java.util.ArrayList;

/**
 * Clase Player: Representa a un jugador en el juego.
 * Esta clase maneja las características, acciones y objetos asociados 
 * a un jugador.
 * Heredera de la clase LabyrinthCharacter.
 */

public class Player extends LabyrinthCharacter{
    
    //*****************************************************************
    // Constantes 
    //*****************************************************************
    
    private static final int MAX_WEAPONS = 2;   // Máximo de armas permitidas
    private static final int MAX_SHIELDS = 3;   // Máximo de escudos permitidos
    private static final int INITIAL_HEALTH = 10;   // Salud inicial
    private static final int HITS2LOSE = 3; // Número de golpes recibidos para derrota
    
    //*****************************************************************
    // Atributos 
    //*****************************************************************
    
    private char number;
    private int consecutiveHits = 0;
    
    private ArrayList <Weapon> weapons; // Lista de armas del jugador
    private ArrayList <Shield> shields; // Lista de escudos del jugador
    
    private ShieldCardDeck shieldCardDeck;  // Cartas de escudo del jugador
    private WeaponCardDeck weaponCardDeck;  // Cartas de arma del jugador

    
    //*****************************************************************
    // Constructores 
    //*****************************************************************
    
    /**
     * Constructor del Player.
     * @param number Número del jugador.
     * @param intelligence Inteligencia defensiva del jugador.
     * @param strength  Fuerza del jugador.
     */
    
    public Player(char number, float intelligence, float strength){ 
        
        super("Player #"+number, intelligence, strength, INITIAL_HEALTH);
        this.number = number;
        weapons = new ArrayList<>();
        shields = new ArrayList<>();
        
        shieldCardDeck = new ShieldCardDeck();
        weaponCardDeck = new WeaponCardDeck();
        
    }
    
    /**
     * Constructor de copia del Player.
     * @param other Jugador a ser copiado.
     */
    
    public Player(Player other){
        
        super(other);    
        this.number = other.number;
        this.weapons = other.weapons; 
        this.shields = other.shields;
        
        this.shieldCardDeck = other.shieldCardDeck;
        this.weaponCardDeck = other.weaponCardDeck;
    }
    
    /**
     * Método para resucitar al jugador.
     * Reinicia las listas de armas y escudos, establece la salud inicial y
     * resetea el contador de hits consecutivos.
     */
    
    public void resurrect()
    {   
        weapons.clear();    // reseteo de armas
        shields.clear();    // reseteo de escudos
        super.setHealth(INITIAL_HEALTH);
        resetHits();
    } 
    
    /**
     * Getter de la fila del jugador.
     * @return La fila en la que se encuentra el jugador.
     */
    
    @Override
    public int getRow()
    {
        return super.getRow();
    }
    
    /**
     * Getter de la columna del jugador.
     * @return La columna donde se encuentra el jugador.
     */
    
    @Override
    public int getCol()
    {
        return super.getCol();
    }
    
    /**
     * Getter del número del jugador.
     * @return El número del jugador actual.
     */
    
    public char getNumber()
    {
        return this.number;
    }
    
    /**
     * Setter de la posición del jugador en el laberinto.
     * @param row
     * @param col 
     */
    
    @Override
    public void setPos(int row, int col)
    {
        super.setPos(row, col);
    }
    
    /**
     * Método que devuelve la dirección de movimiento del jugador.
     * @param direction Dirección de movimiento "deseada"
     * @param validMoves Contenedor de direcciones de movimiento alternativas
     * @return La dirección deseada o la primera en el contenedor de las alternativas
     */
    
    public Directions move(Directions direction, ArrayList<Directions> validMoves)
    {   
        
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        
        if ((size > 0) && (!contained)){
            return validMoves.get(0);
        }
        else{
            return direction;
        }
    }
    
    /**
     * Método que relega su funcionalidad en el método manageHit() para 
     * defenderse de ataques.
     * @param receivedAttack Valor de ataque recibido en coma flotante.
     * @return manageHit()
     */
    
    @Override
    public boolean defend(float receivedAttack)
    {
        return manageHit(receivedAttack);
    }  
    
    /**
     * Método que se encarga de gestionar todo lo relacionado en cuanto a las
     * recompensas recibidas por el jugador cuando gana un combate contra un
     * Monster.
     */
    
    public void receiveReward(){
        
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for (int i = 0; i < wReward; i++){
            Weapon wnew = weaponCardDeck.nextCard();     
            this.receiveWeapon(wnew);
        }
        
        for (int i = 0; i < sReward; i++){
            Shield snew = shieldCardDeck.nextCard();
            receiveShields(snew);   
        }
        
        int extraHealth = Dice.healthReward();
        float playerHealth = getHealth(); 
        setHealth(playerHealth + extraHealth);
        
    }
    
    /**
     * Método toString()
     * @return estado actual, estadísticas, armas y escudos del jugador.
     */
    
    @Override
    public String toString()
    {
        System.out.println("=============================================");
        System.out.println("Armas del jugador #"+number+": "+weapons);
        System.out.println("Escudos del jugador #"+number+": "+shields+"\n");
        return super.toString();
        
    }
    
    /**
     * Método que se encarga de gestionar la recepción de armas como recompensa.
     * @param w Arma aleatoria a recibir por el jugador.
     */
    
    private void receiveWeapon(Weapon w)
    {    
        for (int i = weapons.size()-1; i >= 0; i--){
            boolean discard = weapons.get(i).discard();
            if (discard)
                weapons.remove(weapons.get(i));
        }
        
        int size = weapons.size();
        if (size < MAX_WEAPONS){
            weapons.add(w);
        }
    }
    
    /**
     * Método que se encarga de gestionar la recepción de escudos como recompensa.
     * @param s Escudo aleatorio a recibir por el jugador.
     */
    
    private void receiveShields(Shield s)
    {
        for(int i = shields.size()-1; i >= 0; i--){
            boolean discard = shields.get(i).discard();
            if(discard)
                shields.remove(shields.get(i));
        }
        
        int size = shields.size();
        if(size < MAX_SHIELDS)
            shields.add(s);
    }
    
 /*   Métodos no usados pero que figuran en el UML del proyecto
    
    private Weapon newWeapon()
    {
        Weapon newWeapon = new Weapon(Dice.weaponPower(),Dice.usesLeft());
        weapons.add(newWeapon);
        return newWeapon;
    }
    
    // Método newShield()
    // Igual que el método newWeapon, pero llamando a la clase Shield
    
    private Shield newShield()
    {
        Shield newShield = new Shield(Dice.shieldPower(), Dice.usesLeft());
        shields.add(newShield);
        return newShield;
        
    }
*/
    
    /**
     * Método que devuelve la suma de poder de ataque de todas las armas del
     * jugador.
     * @return La suma total de ataque.
     */
    
    protected float sumWeapons()   // idea
    {
        float sum = 0f;
        
        for (int i = 0; i < weapons.size(); i++){
            sum += weapons.get(i).attack();
        }
        
        return sum;
                
    }
    
    /**
     * Método que devuelve la suma de poder de protección de todos los escudos
     * del jugador.
     * @return La suma total de protección.
     */
    
    protected float sumShields() 
    {
        float sum = 0f;
        
        for (int i = 0; i < shields.size(); i++){
            sum += shields.get(i).protect();
        }
        
        return sum;
    }
    
    /**
     * Método que devuelve el poder total de ataque del jugador, suma de su
     * fuerza con la suma de todas sus armas disponibles en la lista.
     * @return Poder total de ataque del jugador.
     */
    
    @Override
    public float attack(){
        return super.getStrength() + sumWeapons();
    }
    
    /**
     * Método que devuelve el poder total de defensa del jugador, suma de su
     * inteligencia defensiva con la suma de todos sus escudos disponibles en
     * la lista.
     * @return Poder total de defensa del jugador.
     */
    
    protected float defensiveEnergy()
    {
        return super.getIntelligence() + sumShields();
    }
    
    /**
     * Método que gestiona la defensa del jugador a los ataques de un monstruo.
     * @param receivedAttack ataque recibido.
     * @return True si ha perdido el combate, false si lo ha ganado.
     */
    
    public boolean manageHit(float receivedAttack)
    {
        float defense = defensiveEnergy();
        if (defense < receivedAttack) {
            super.gotWounded();
            incConsecutiveHits();
        }
        else {
            resetHits();
        }
       
        if ((consecutiveHits == HITS2LOSE) || (super.dead() )) {
            resetHits();
            return true;
        }
        else {
            return false;
        }
    }
    
    /**
     * Método que resetea los hits consecutivos.
     */
    
    public void resetHits(){
        this.consecutiveHits = 0;
    }
    
    /**
     * Método que incrementa los hits consecutivos de uno en uno.
     */
    
    public void incConsecutiveHits(){
        this.consecutiveHits++;
    }
    
    
}

File: Irrgarten-Ruby/irrgarten/player.rb
#encoding:utf-8

=begin

La clase Player es una subclase de LabyrinthCharacter dentro del módulo Irrgarten.
Representa a un jugador dentro de un laberinto y tiene atributos y métodos para administrar
sus características, movimientos, ataques, defensa y gestión de objetos como armas y escudos.

=end

require_relative 'dice'
require_relative 'shield'
require_relative 'weapon'
require_relative 'directions'
require_relative 'labyrinth_character'

module Irrgarten

  class Player < LabyrinthCharacter

	 @@MAX_WEAPONS = 2
	 @@MAX_SHIELDS = 3
	 @@INITIAL_HEALTH = 10
	 @@HITS2LOSE = 3

	 attr_reader :consecutive_hits, :number, :weapons, :shields

	 # Constructor
	 def initialize(number, intelligence, strength)
		super("Player ##{number}", intelligence.to_f, strength.to_f, @@INITIAL_HEALTH)

		#Inicialización del resto de atributos con valores por defecto
		@number = number	# Número identificador del jugador
		@row = 0	#Número de golpes consecutivos recibidos
		@col = 0
		@consecutive_hits = 0

		#Arrays de armas y escudos
		@weapons = []
		@shields = []

	 end

	 #Constructor de copia
	 def other_player(other)
		other_lab_character(other)
		@number = other.number
		@consecutive_hits = 0

		@shields = other.shields
		@weapons = other.weapons
	 end

	 # Reinicia los atributos y contenedores de un jugador al resucitar
	 def resurrect
		@weapons.clear
		@shields.clear
		@health = @@INITIAL_HEALTH.to_f
		@consecutive_hits = 0
	 end

	 # Fila del jugador
	 def row
		super
	 end

	 # Columna del jugador
	 def col
		super
	 end

	 # Obtiene el número del jugador
	 def get_number
		@number
	 end

	 # Establece la posición del jugador
	 def set_pos(row, col)
		super
	 end

	 # Indica si el jugador está muerto
	 def dead
		super
	 end

	 # Mueve al jugador en una dirección válida
	 def move(direction, valid_moves)
		size = valid_moves.size
		contained = valid_moves.include?(direction)

		if size > 0 && !contained
		  return valid_moves[0]
		else
		  return direction
		end

	 end

	 # Calcula el poder de ataque del jugador
	 def attack
		sum_weapons + @strength
	 end

	 # Calcula el poder defensivo del jugador
	 def defend(received_attack)
		manage_hit(received_attack)
	 end

	 # Otorga recompensas al jugador como armas, escudos y salud extra
	 def receive_reward
		w_reward = Dice.weapons_reward
		s_reward = Dice.shields_reward

		(0..w_reward).each do |i|
		  wnew = new_weapon
		  receive_weapon(wnew)
		end

		(0..s_reward).each do |i|
		  snew = new_shield
		  receive_shield(snew)
		end

		extra_health = Dice.health_reward
		@health += extra_health

	 end

	 # ToString
	 def to_s
		"Player" << super << "\nArmas: #{@weapons.to_s}\nEscudos: #{@shields.to_s}"
	 end

	 # Nombre del jugador
	 def name
		@name
	 end

	 private

	 # Método encargado de gestionar la recepción de recompensas de tipo Weapon
	 def receive_weapon(w)
		@weapons.each do |i|
		  discard = i.discard
		  if discard
			 @weapons.delete(i)
		  end
		end

		size = @weapons.size
		if size < @@MAX_WEAPONS
		  @weapons << w
		end

	 end

	 # Método encargado de gestionar la recepción de recompensas de tipo Escudo
	 def receive_shield(s)
		@shields.each do |i|
		  discard = i.discard
		  if discard
			 @shields.delete(i)
		  end
		end

		size = @shields.size
		if size < @@MAX_SHIELDS
		  @shields << s
		end

	 end

	 # Crear un nuevo arma
	 def new_weapon
		Weapon.new(Dice.weapon_power, Dice.uses_left)
	 end

	 # Crear un nuevo escudo
	 def new_shield
		Shield.new(Dice.shield_power, Dice.uses_left)
	 end

	 # Calcula la suma de ataque de todas las armas de jugador
	 def sum_weapons
		sum = 0.0
		@weapons.each do |x|
		  sum += x.attack
		end

		sum
	 end

	 # Calcula la suma de protección de los escudos del jugador
	 def sum_shields
		sum = 0.0
		@shields.each do |x|
		  sum += x.protect
		end

		sum
	 end

	 # Calcula la energía defensiva del jugador
	 def defensive_energy
		@intelligence + sum_shields
	 end

	 # Gestiona la recepción de golpes por parte de un monstruo
	 def manage_hit(received_attack)
		defense = defensive_energy
		if defense < received_attack
		  got_wounded
		  inc_consecutive_hits
		else
		  reset_hits
		end

		if @consecutive_hits == @@HITS2LOSE || dead
		  reset_hits
		  return true
		else
		  return false
		end

	 end

	 # Resetea el contador de hits a cero
	 def reset_hits
		@consecutive_hits = 0
	 end

	 # Reduce la salud del jugador en 1 unidad
	 def got_wounded
		super
	 end

	 # Incrementa el contador de hits consecutivos en 1
	 def inc_consecutive_hits
		@consecutive_hits += 1
	 end

  end #class

end	#module

